compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1209,21 +1209,28 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                         throw_ub_custom!(fluent::const_eval_copy_nonoverlapping_overlapping);
                     }
                 }
+            }
 
-                for i in 0..num_copies {
-                    ptr::copy(
-                        src_bytes,
-                        dest_bytes.add((size * i).bytes_usize()), // `Size` multiplication
-                        size.bytes_usize(),
-                    );
+            let size_in_bytes = size.bytes_usize();
+            // For particularly large arrays (where this is perf-sensitive) it's common that
+            // we're writing a single byte repeatedly. So, optimize that case to a memset.
+            if size_in_bytes == 1 && num_copies >= 1 {
```

Comment:
At this point we know both the src and dest ranges are non-empty. So the `num_copies` check is unnecessary.

Reply:
I wasn't able the quickly convince myself the above logic guarantees that, but happy to remove it if so. Feels somewhat non-loval as a guarantee though unless I'm missing something.

Reply:
It would be very problematic if this code was reachable for 0-size dst. After all, here in the code we have an `AllocRefMut`, which implies there is a backing allocation, but when dst is zero-sized then we do not even require a backing allocation, so we must bail out earlier.

All our memory access logic works that way; that's why `get_ptr_access` distinguishes zero-sized from non-zero-sized.

Reply:
Oh, actually... looks like I may have spoken a bit too quickly and assumed https://github.com/rust-lang/rust/issues/117945 was already implemented... right now `check_and_deref_ptr` can still return `Some` even for size 0. oopsie

Reply:
Actually no they can't.^^ The doc comment is just outdated.